python/cudf/bindings/dlpack.pyx

```diff
@@ -0,0 +1,126 @@
+                  " order) input. If the input tensor is row-major, transpose"
+                  " it before passing it to this function.")
+
+    cdef DLManagedTensor* dlpack_tensor = <DLManagedTensor*>pycapsule.PyCapsule_GetPointer(dlpack_capsule, 'dltensor')
+    pycapsule.PyCapsule_SetName(dlpack_capsule, 'used_dltensor')
+    cdef gdf_size_type c_result_num_cols
+    cdef gdf_column* result_cols
+
+    with nogil:
+        result = gdf_from_dlpack(
+            &result_cols,
+            &c_result_num_cols,
+            dlpack_tensor
+        )
+
+    check_gdf_error(result)
+    
+    # TODO: Replace this with a generic function from cudf_cpp.pyx since this
+    # is copied from join.pyx
+    res = []
+    valids = []
+
+    for idx in range(c_result_num_cols):
+        data_ptr = <uintptr_t>result_cols[idx].data
+        if data_ptr:
+            res.append(
+                rmm.device_array_from_ptr(
+                    ptr=data_ptr,
+                    nelem=result_cols[idx].size,
+                    dtype=gdf_to_np_dtype( result_cols[idx].dtype),
+                    finalizer=rmm._make_finalizer(data_ptr, 0)
+                )
+            )
+        else:
+            res.append(
+                rmm.device_array(
+                    0,
+                    dtype=gdf_to_np_dtype( result_cols[idx].dtype)
+                )
+            )
+
+        valid_ptr = <uintptr_t>result_cols[idx].valid
+        if valid_ptr:
+            valids.append(
+                rmm.device_array_from_ptr(
+                    ptr=valid_ptr,
+                    nelem=calc_chunk_size(result_cols[idx].size, mask_bitsize),
+                    dtype=mask_dtype,
+                    finalizer=rmm._make_finalizer(valid_ptr, 0)
+                )
+            )
+        else:
+            valids.append(None)
+
+    return res, valids
+
+
+cpdef to_dlpack(in_cols):
+    """
+    Converts a a list of cudf Column objects into a DLPack Tensor PyCapsule.
+    """
+
+    warnings.warn("WARNING: cuDF to_dlpack() produces column-major (Fortran "
+                  "order) output. If the output tensor needs to be row major, "
+                  "transpose the output of this function.")
+
+    input_num_cols = len(in_cols)
+
+    cdef DLManagedTensor* dlpack_tensor =<DLManagedTensor*>malloc(sizeof(DLManagedTensor))
+    cdef gdf_column** input_cols = <gdf_column**>malloc(input_num_cols * sizeof(gdf_column*))
+    cdef gdf_size_type c_input_num_cols = input_num_cols
+
+    for idx, col in enumerate(in_cols):
+        check_gdf_compatibility(col)
+        input_cols[idx] = column_view_from_column(col)
+
+    with nogil:
+        result = gdf_to_dlpack(
+            dlpack_tensor,
+            input_cols,
+            c_input_num_cols
+        )
+
+    check_gdf_error(result)
+
+    return pycapsule.PyCapsule_New(dlpack_tensor, 'dltensor', pycapsule_deleter)
+
+cdef void pycapsule_deleter(object pycap_obj):
```

Comment:
Suggest a more specific function name here like `dlmanaged_tensor_pycapsule_deleter`. I don't think that `pycapsule_deleter` is accurate.

Reply:
```suggestion
cdef void dlmanaged_tensor_pycapsule_deleter(object pycap_obj):
```